v1/remote/image.go

```diff
@@ -16,19 +16,145 @@ package remote
 
 import (
 	"fmt"
+	"io"
 	"net/http"
+	"net/url"
+	"sync"
 
 	"github.com/google/go-containerregistry/authn"
 	"github.com/google/go-containerregistry/name"
 	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/partial"
+	"github.com/google/go-containerregistry/v1/remote/transport"
+	"github.com/google/go-containerregistry/v1/types"
 )
 
-// TODO(jonjohnson): This.
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+// image accesses an image from a remote registry
+type image struct {
+	ref    name.Reference
+	client *http.Client
+}
+
+type compressedImage struct {
+	*image
+	manifestLock sync.Mutex // Protects manifest
+	manifest     *v1.Manifest
+	configLock   sync.Mutex // Protects config
+	config       *v1.ConfigFile
+}
+
+var _ partial.CompressedImageCore = (*compressedImage)(nil)
 
 // Image accesses a given image reference over the provided transport, with the provided authentication.
 func Image(ref name.Reference, auth authn.Authenticator, t http.RoundTripper) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: remote.Image(%v)", ref)
+	tr, err := transport.New(ref, auth, t, transport.PullScope)
+	if err != nil {
+		return nil, err
+	}
+	img := image{
+		ref: ref,
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+	c := compressedImage{
+		image: &img,
+	}
+	return partial.CompressedToImage(&c)
+}
+
+func (i *image) url(resource, identifier string) url.URL {
+	return url.URL{
+		Scheme: transport.Scheme(i.ref.Context().Registry),
+		Host:   i.ref.Context().RegistryStr(),
+		Path:   fmt.Sprintf("/v2/%s/%s/%s", i.ref.Context().RepositoryStr(), resource, identifier),
+	}
+}
+
+// TODO(jonjohnson): Handle manifest lists.
+// TODO(jonjohnson): Validate content matches Docker-Image-Digest.
+func (c *compressedImage) Manifest() (*v1.Manifest, error) {
+	c.manifestLock.Lock()
+	defer c.manifestLock.Unlock()
+	if c.manifest != nil {
+		return c.manifest, nil
```

Comment:
Thinking about this, we're going to need a `RawManifest` to support copying images preserving their digest.

I don't think any action is needed, but something of which to be aware.